activerecord/lib/active_record/attributes.rb

```diff
@@ -245,8 +246,17 @@ def define_attribute(
 
       def load_schema! # :nodoc:
         super
-        attributes_to_define_after_schema_loads.each do |name, (type, options)|
-          define_attribute(name, _lookup_cast_type(name, type, options), **options.slice(:default))
+        attributes_to_define_after_schema_loads.each do |name, (type, options, transform)|
+          if type.is_a?(Symbol)
+            adapter_name = ActiveRecord::Type.adapter_name_from(self)
+            type = ActiveRecord::Type.lookup(type, **options.except(:default), adapter: adapter_name)
+          elsif type.nil?
+            type = type_for_attribute(name)
+          end
+
+          type = transform[type] if transform
```

Comment:
This feature (specific type with decoration `attribute :type do |type| ... end`) is intentionally limited since it is not internally used.

If we'd like to provide internally unused feature, I don't think it's appropriate to sneak it in as part of the refactoring.

Reply:
Ah, I thought the block of `attribute` was meant to be an internal-only feature, since it is undocumented.  Would it be good to make it an internal-only feature?  If so, should we rename it to `_block`?

But if not, I can enforce the limitation by raising an `ArgumentError` when both `cast_type` and block are specified.

Reply:
> But if not, I can enforce the limitation by raising an `ArgumentError` when both `cast_type` and block are specified.

That sounds right to me 👍

Reply:
Ah, looks like this is actually used internally for the `EnumType`. Not sure how to do that in any other way without the block form? I guess you'd end up in awkward `type_for_attribute` before then calling `attribute`, doesn't sound right.